### PR TITLE
Rework arg bundles for begin and on functions

### PR DESCRIPTION
**Pull request by mppf**
Work I have been doing on string-as-rec investigates using a
serialize/deserialize interface for argument bundles (#2735).

Partly in support of that string-as-rec investigation, this patch separates the
many uses of the class containing the bundled arguments into different explicit
arguments for task/on functions. I believe that this represents a worthwhile
clean-up no matter what happens with the serialize/deserialize patches.

Previous to this patch, there was some pretty nasty code run during code
generation to find the taskList in an EndCount during task spawn.

For task body functions that eventually call 
chpl_taskListAddBegin/chpl_taskListAddCoStmt:

  Before this patch:

    // get(1) is an class instance for bundled arguments

  After this patch:

    // get(1) is a ref/wide ref to a task list value
    // get(2) is the node ID owning the task list
    // get(3) is a buffer containing bundled arguments
    // get(4) is the buffer's length (currently unused for task fns)
    // get(5) is a dummy class type for the argument bundle

For on body functions that eventually call
chpl_executeOn/chpl_executeOnNB/chpl_executeOnFast:

  Before this patch:

    // get(1) is the locale
    // get(2) is a class instance for bundled arguments

  After this patch:

    // get(1) is the locale
    // get(2) is a buffer containing bundled arguments
    // get(3) is a the size of the buffer
    // get(4) is a dummy class type for the argument bundle

This patch makes the following specific changes:
 * adds dtCVoidPtr in order to support compiler interaction
   with C functions expecting a buffer
 * moves discovery of the task list from code generation (expr.cpp)
   to parallel.cpp.
 * adds CallExpr constructors with 5 arguments to support this change
 * update insertLineNumbers and insertWideReferences to work with
   the new way to call a task/on function.

This patch replaces #2447.

- [x] full local testing
- [x] full gasnet testing

### CODE REVIEW

**Comment by mppf** [comment]
Local testing has many tests failing with TYP1250...

**Comment by mppf** [comment]
OK, local testing is good. Running GASNet testing now.

**Comment by mppf** [comment]
Gasnet testing underway...

**Comment by mppf** [comment]
GASNet testing showed these 7 failures:
[Error: Timed out executing program distributions/robust/arithmetic/performance/multilocale/alloc_all]
[Error matching compiler output for multilocale/diten/needMultiLocales/DijkstraTermination]
[Error matching compiler output for optimizations/cache-remote/ferguson/correctness/acqrel-sync-subtask]
[Error matching compiler output for parallel/taskPar/sungeun/taskCount (compopts: 1)]
[Error matching compiler output for parallel/taskPar/sungeun/taskCount (compopts: 2)]
[Error matching compiler output for performance/sungeun/multilocale/syncsingle]
[Error matching compiler output for users/vass/domain_with_on-all]

**Comment by mppf** [comment]
Passed the failing tests. Passed GASNet test/multilocale.

File: compiler/passes/parallel.cpp [hunk 1]
```diff
@@ -322,9 +322,93 @@ bundleArgs(CallExpr* fcall, BundleArgsFnData &baData) {
...
+  VarSymbol *taskList = NULL;
+  VarSymbol *taskListNode = NULL;
+
+  if (!fn->hasFlag(FLAG_ON)) {
+    for_actuals(arg, fcall) {
+      bool strcmp_found = false;
+      bool type_found = false;
+
+      Type* baseType = arg->getValType();
+      if (baseType->symbol->hasFlag(FLAG_END_COUNT)) {
```
Comment (benharsh) [inline compiler/passes/parallel.cpp]: It doesn't look like FLAG_END_COUNT is ever added to anything. Is it worthwhile to include this check in this PR?

Comment (benharsh) [inline compiler/passes/parallel.cpp]: Whoops, I totally missed the pragma in the internal modules, nevermind.

File: compiler/passes/parallel.cpp [hunk 2]
```diff
@@ -322,9 +322,93 @@ bundleArgs(CallExpr* fcall, BundleArgsFnData &baData) {
...
+      }
+
+      // This strcmp code was moved from expr.cpp codegen,
+      // but there has got to be a better way to do this! 
+      if (strstr(baseType->symbol->name, "_EndCount") != NULL) {
+        strcmp_found = true;
+      }
+
+      // If this assert never fires, we can remove the strcmp version.
+      INT_ASSERT(type_found == strcmp_found);
```
Comment (benharsh) [inline compiler/passes/parallel.cpp]: Do you plan on removing the strstr if nightly testing indicates that it is unnecessary?

**Comment by benharsh** [comment]
Otherwise, I think this looks great! :+1: :+1:

File: compiler/passes/parallel.cpp (hunk 2 above)
Comment (mppf) [inline compiler/passes/parallel.cpp]: Yes, that's the idea.

**Pull request merged**